ios/RNMagicScript/components/NodesManager/NodesManager.swift

```diff
@@ -0,0 +1,217 @@
+//
+//  Copyright (c) 2019 Magic Leap, Inc. All Rights Reserved
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import SceneKit
+import ARKit
+
+//sourcery: AutoMockable
+//sourcery: ObjcProtocol
+@objc protocol NodesManaging: NSObjectProtocol {
+    @objc var scene: Scene? { get }
+    @objc var prismsById: [String: Prism] { get }
+
+    @objc func registerScene(_ scene: Scene, sceneId: String)
+    @objc func registerPrism(_ prism: Prism, prismId: String)
+    @objc func registerNode(_ node: TransformNode, nodeId: String)
+    @objc func unregisterNode(_ nodeId: String)
+
+    @objc func updateNode(_ nodeId: String, properties: [String: Any]) -> Bool
+}
+
+@objc open class NodesManager: NSObject {
+    @objc public static let instance = NodesManager(rootNode: BaseNode(), nodesById: [:], transformNodeByAnchorUuid: [:], anchorNodeByAnchorUuid: [:])
+    @objc public private (set) var rootNode: BaseNode
+
+    fileprivate var nodesById: [String: BaseNode]
+    fileprivate var transformNodeByAnchorUuid: [String: TransformNode]
+    fileprivate var anchorNodeByAnchorUuid: [String: SCNNode]
+
+    fileprivate(set) var nodeSelector: UiNodeSelector!
+    fileprivate weak var ARView: RCTARView?
+    var dialogPresenter: DialogPresenting?
+
+    init(rootNode: BaseNode, nodesById: [String: TransformNode], transformNodeByAnchorUuid: [String: TransformNode], anchorNodeByAnchorUuid: [String: SCNNode]) {
+        self.rootNode = rootNode
+        self.nodesById = nodesById
+        self.transformNodeByAnchorUuid = transformNodeByAnchorUuid
+        self.anchorNodeByAnchorUuid = anchorNodeByAnchorUuid
+    }
+
+    @objc public func registerARView(_ arView: RCTARView) {
+        ARView = arView
+        arView.scene.rootNode.addChildNode(rootNode)
+        arView.register(self)
+        nodeSelector = UiNodeSelector(rootNode)
+    }
+
+    @objc public func findNodeWithId(_ nodeId: String) -> BaseNode? {
+        return nodesById[nodeId]
+    }
+
+    @objc public func findUiNodeWithId(_ nodeId: String) -> UiNode? {
+        return nodesById[nodeId] as? UiNode
+    }
+
+    @objc public func findNodeWithAnchorUuid(_ nodeId: String) -> TransformNode? {
+        return transformNodeByAnchorUuid[nodeId]
+    }
+
+    @objc public private(set) var scene: Scene?
+    @objc public func registerScene(_ scene: Scene, sceneId: String) {
+        scene.name = sceneId
+        self.scene = scene
```

Comment:
Should we unregister the previous scene first? Should we use assert here to forbid registering multiple scenes?